chainercv/datasets/between_class_dataset.py

```diff
@@ -0,0 +1,104 @@
+import random
+
+import numpy
+
+from chainer.dataset import dataset_mixin
+
+
+class BetweenClassLabeledImageDataset(dataset_mixin.DatasetMixin):
+
+    """Dataset which returns mixed images and labels for bc learning[1].
+
+    `BetweenClassLabeledImageDatasets` chooses two images which labels are
+    different from base dataset, and mix these images and labels respectively.
+
+    Two mix methods for images are prepared. One is 'simple', which simply
+    calculate weighted average. The other is 'bc_plus', which mixes images as
+    the waveform data. Apart from these two methods, the users can define and
+    use the original mix method.
+    In all cases, weight is chosen in [0, 1).
+
+    Since the combination of images is chosen randomly, the number of times an
+    image is used for each epoch is different.
+
+    Unlike `LabeledImageDatasets`, label is a one-dimensional float array with
+    two nonzero weights. The summed weights is one.
+
+    The base dataset `__getitem__` should return image and label. Please see
+    the following example.
+
+    >>> from chainer.datasets import get_mnist
+    >>> from chainer.datasets import TransformDataset
+    >>> from chainercv.datasets import BetweenClassLabeledImageDataset
+    >>> dataset, _ = get_mnist()
+    >>> def transform(in_data):
+    ...     img, label = in_data
+    ...     img -= 0.5  # scale to [-0.5, -0.5]
+    ...     return img, label
+    >>> dataset = TransformDataset(dataset, transform)
+    >>> dataset = BetweenClassLabeledImageDataset(dataset)
+
+    Args:
+        dataset: The underlying dataset. The index of this dataset corresponds
+            to the index of the base dataset. This object needs to support
+            functions :meth:`__getitem__` and :meth:`__len__` as described
+            above.
+        dtype: Data type of resulting image arrays.
+        label_dtype: Data type of the labels.
+        method (callable or 'simple' or 'bc_plus'): If callable, the result of
+            two image mixture is `method(image1, image2, mix_ratio)`. Else,
+            Perform calculations according to the original paper. Default is
+            'bc_plus'.
+
+    See the papers for details: [1]`Between-class Learning for Image \
+    Classification <https://arxiv.org/abs/1711.10284>`_.
+
+    """
+
+    def __init__(self, dataset, dtype=numpy.float32, label_dtype=numpy.float32,
```

Comment:
Could you remove `dtype` and `label_dtype` from arguments?
By default, images and soft-labels are `dtype=np.float32`.
If different dtypes are needed, transforms can be used.